src/coap_debug.c

```diff
@@ -189,37 +197,44 @@ coap_print_addr(const struct coap_address_t *addr, unsigned char *buf, size_t le
 
     break;
   default:
-    memcpy(buf, "(unknown address type)", min(22, len));
+    /* Include trailing NULL if possible */
+    memcpy(buf, "(unknown address type)", min(22+1, len));
+    buf[len-1] = '\000';
     return min(22, len);
   }
 
   /* Cast needed for Windows, since it doesn't have the correct API signature. */
   if (inet_ntop(addr->addr.sa.sa_family, addrptr, (char *)p,
                 min(len, need_buf)) == 0) {
     perror("coap_print_addr");
+    *p = '\000';
     return 0;
   }
 
-  p += strnlen((char *)p, len);
+  p += strlen((char *)p);
 
   if (addr->addr.sa.sa_family == AF_INET6) {
-    if (p < buf + len) {
+    if (p < buf + len + 1) {
       *p++ = ']';
     } else
-      return 0;
+      return p - buf; /* Already NULL terminated */
   }
 
-  p += snprintf((char *)p, buf + len - p + 1, ":%d", port);
+  /* Cannot rely on snprintf() return value for short buffers */
+  snprintf((char *)p, buf + len - p - 1, ":%d", port);
```

Comment:
As snprintf adds a terminating zero, the length should be `buf + len - p`.

Reply:
Fixed